mainwindow.cpp

```diff
@@ -0,0 +1,334 @@
+        msg += R"*("!)*";
+
+        QMessageBox::critical(this, "Invalid email detected!", msg);
+        return false;
+    }
+    return true;
+}
+
+void MainWindow::SpawnNewHistoryUnit(const LetterStruct& Letter)
+{
+    QVBoxLayout* layout = qobject_cast<QVBoxLayout*>(ui->MailHistoryScrollArea->layout());
+    MailHistoryUnit* new_history_unit = nullptr;
+
+    if (layout)
+    {
+        new_history_unit = new MailHistoryUnit(Letter);
+        layout->setAlignment(Qt::AlignHCenter | Qt::AlignTop);
+        layout->insertWidget(0, new_history_unit);
+    }
+
+    connect(new_history_unit, SIGNAL(OnMouseReleased(QVector<LetterStruct>)), this, SLOT(HistoryWidgetClicked(QVector<LetterStruct>)));
+}
+
+void MainWindow::SpawnNewHistoryUnit(const QVector<LetterStruct>& Letters)
+{
+    QVBoxLayout* layout = qobject_cast<QVBoxLayout*>(ui->MailHistoryScrollArea->layout());
+    MailHistoryUnit* new_history_unit = new MailHistoryUnit(Letters);
+
+    if (layout)
+    {
+        new_history_unit = new MailHistoryUnit(Letters);
+        layout->setAlignment(Qt::AlignHCenter | Qt::AlignTop);
+        layout->insertWidget(0, new_history_unit);
+    }
+
+    connect(new_history_unit, SIGNAL(OnMouseReleased(QVector<LetterStruct>)), this, SLOT(HistoryWidgetClicked(QVector<LetterStruct>)));
+}
+
+void MainWindow::PopulateMailsHistory()
+{
+    QDir dir(m_temp_file_path);
+
+    if (!dir.exists())
+    {
+        qDebug() << "Directory does not exist:" << m_temp_file_path;
+    }
+
+    dir.setSorting(QDir::Time);
+    dir.setFilter(QDir::Files);
+    QStringList files = dir.entryList();
+
+    for (const QString &fileName : files)
+    {
+        qDebug() << fileName;
+        QVector<LetterStruct> letters = ReadLettersFromFile(m_temp_file_path + fileName);
+        if (!letters.isEmpty()) SpawnNewHistoryUnit(letters);
+    }
+}
+
+void MainWindow::SendLetter(const LetterStruct& Letter)
+{
+
+}
+
+bool MainWindow::WriteLettersToFile(const QVector<LetterStruct>& letters, const QString& full_file_name)
+{
+    QFile file(full_file_name);
+    if (!file.open(QIODevice::ReadWrite))
+    {
+        qDebug() << "Could not open file for writing";
+        return false;
+    }
+    if (letters.isEmpty())
+    {
+        qDebug() << "Trying to write 0 letters!";
+        return false;
+    }
+
+    qDebug() << "Writing to " << full_file_name << " " << letters.size() << " letters";
+
+    QDataStream out(&file);
+    out.setVersion(QDataStream::Qt_6_7);
+
+    out << letters.size();
+
+    for (const auto& letter : letters)
+    {
+        out << letter;
+    }
+
+    file.close();
+
+    return true;
+}
+
+QVector<LetterStruct> MainWindow::ReadLettersFromFile(const QString& full_file_name)
+{
+    QFile file(full_file_name);
+    QVector<LetterStruct> Letters;
+
+    if (!file.open(QIODevice::ReadOnly))
+    {
+        qDebug() << "Could not open file " << full_file_name << " for reading";
+        return Letters;
+    }
+    if (!QFile::exists(full_file_name))
```

Comment:
you are already checking the existence of full_file_name when trying to open it; checking one more time is not nessesary

Reply:
thanks for noticing, will be fixed!